include/envoy/server/lifecycle_notifier.h

```diff
@@ -29,14 +29,22 @@ class ServerLifecycleNotifier {
 
   /**
    * Callback invoked when the server reaches a certain lifecycle stage.
+   *
+   * Instances of the second type which take an Event::PostCb parameter must post
+   * that callback to the main dispatcher when they have finished processing of
+   * the new lifecycle state. This is useful when the main dispatcher needs to
```

Comment:
I'd be explicit in this comment that server stage processing will be delayed until that callback is posted.  I presume that works both for delaying startup and delaying shutdown?

Reply:
Whether the server will wait for callbacks to complete depends on the stage - currently the server waits during shutdown but not at startup. I added comments about this.

Reply:
Up to you, but in my ideal world the behavior would be the same for both changes.  If you can submit a callback with cb argument for the Startup, I feel like that behavior should be implemented.  (Alternatively, you could assert not implemented or throw an exception or something if someone tries to do it.)  It just feels funny to leave a tool around that looks like it should do something, but won't, especially when the difference between it doing something and not doing something might be covered up by a race (i.e. someone could use it and it looks like it's working but it doesn't).